File: crates/harness-subagents/src/claude_code.rs
//! `subagent.claude_code` — delegate a coding task to the local
//! `claude` CLI (Anthropic's Claude Code distribution). Spawns
//! `claude --print --output-format stream-json --verbose` with the
//! task as a positional argument, parses each `SDKMessage` line on
//! stdout, and emits the equivalent [`SubAgentEvent`] frames so the
//! outer UI can show Claude Code's reasoning + tool calls in real
//! time alongside the rest of the subagent surface.
//!
//! Why CLI and not the Node SDK? Most operators who care about
//! `subagent.claude_code` already have the `claude` binary
//! installed (it's the canonical distribution); requiring an extra
//! `npm i -g @anthropic-ai/claude-agent-sdk` step on top doubles
//! the install friction. The CLI's `stream-json` format is also
//! the SDK's wire format on stdout, so we re-use the same
//! `SDKMessage` shape either way — switching to the binary just
//! drops the Node sidecar and its bundled script.
//!
//! Wire protocol (`claude --print --output-format stream-json --verbose`):
//!
//! ```text
//!   { "type": "system",    "subtype": "init", ... }
//!   { "type": "assistant", "message": { "content": [{ "type": "text", "text": "..." }] }, ... }
//!   { "type": "assistant", "message": { "content": [{ "type": "tool_use", "id": "...", "name": "Read", "input": { ... } }] }, ... }
//!   { "type": "user",      "message": { "content": [{ "type": "tool_result", "tool_use_id": "...", "content": "..." }] }, ... }
//!   { "type": "result",    "subtype": "success", "result": "...", "is_error": false, ... }
//! ```
//!
//! Discovery: the composition root calls [`probe`] once at startup;
//! if `claude` isn't on PATH (or `--version` fails), it returns
//! `Err(reason)` and the subagent is **not registered** — no panic,
//! no hard error. Mirrors how `harness-mcp` skips servers it can't
//! launch.
//!
//! Authentication: the spawned `claude` inherits the parent
//! process's env, so a user who has logged in interactively
//! (`claude /login`) gets seamless auth — the credentials live in
//! the OS keychain (or `~/.claude/.credentials.json`). If keychain
//! access is unavailable, set `ANTHROPIC_API_KEY` for the
//! `jarvis` process and add `--bare` via
//! [`ClaudeCodeConfig::extra_args`] (TODO).

use crate::{Artifact, SubAgent, SubAgentEvent, SubAgentFrame, SubAgentInput, SubAgentOutput};
use async_trait::async_trait;
use harness_core::{emit_subagent, BoxError};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tracing::{debug, warn};

pub const DESCRIPTION: &str = "Delegate a coding task to Claude Code (Anthropic's `claude` CLI). It can read, edit, and run shell commands inside the workspace; treat it like a peer coder you hand a focused task to. Will modify files; the call is approval-gated.";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClaudeCodeConfig {
    /// Path or name of the `claude` binary. Default `claude` (PATH
    /// lookup). Override via `JARVIS_SUBAGENT_CLAUDE_CODE_BIN` at
    /// the composition root.
    pub claude_bin: String,
    /// Optional model override forwarded to the CLI as `--model`.
    /// `None` lets the CLI use its configured default (whatever the
    /// user picked via `claude /model` or their settings).
    pub model: Option<String>,
}

impl Default for ClaudeCodeConfig {
    fn default() -> Self {
        Self {
            claude_bin: "claude".into(),
            model: None,
        }
    }
}

pub struct ClaudeCodeSubAgent {
    config: ClaudeCodeConfig,
}

impl ClaudeCodeSubAgent {
    pub fn new(config: ClaudeCodeConfig) -> Self {
        Self { config }
    }
}

/// Probe the host: is the `claude` binary on PATH and does
/// `--version` recognise it as Claude Code? Run once at startup;
/// on failure callers should skip registering the subagent (a
/// one-line INFO log is the right level).
pub async fn probe(claude_bin: &str) -> Result<(), String> {
    let output = Command::new(claude_bin)
        .arg("--version")
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .stdin(Stdio::null())
        .output()
        .await
        .map_err(|e| format!("spawn `{claude_bin}` failed: {e}"))?;
    if !output.status.success() {
        return Err(format!(
            "`{claude_bin} --version` exited with {}",
            output.status
        ));
    }
    let version = String::from_utf8_lossy(&output.stdout);
    // The CLI prints e.g. `2.1.119 (Claude Code)` — the parenthesised
    // tag is what disambiguates the official binary from a someone-
    // else's `claude` on PATH.
    if !version.contains("Claude Code") {
        return Err(format!(
            "`{claude_bin} --version` output unrecognised (got: {})",
            version.trim()
        ));
    }
    Ok(())
}

#[async_trait]
impl SubAgent for ClaudeCodeSubAgent {
    fn name(&self) -> &str {
        "claude_code"
    }
    fn description(&self) -> &str {
        DESCRIPTION
    }
    fn input_schema(&self) -> serde_json::Value {
        serde_json::json!({
            "type": "object",
            "properties": {
                "task": {
                    "type": "string",
                    "description": "Natural-language coding task. Claude Code will plan + execute autonomously."
                }
            },
            "required": ["task"],
            "additionalProperties": false
        })
    }
    fn requires_approval(&self) -> bool {
        // The CLI *will* mutate the workspace. The outer approver
        // MUST gate this call; per-tool approvals inside the CLI are
        // governed by `--permission-mode bypassPermissions` (the
        // user already approved the outer call, no point asking
        // again — and `--print` mode can't surface interactive
        // prompts anyway).
        true
    }

    async fn invoke(&self, input: SubAgentInput) -> Result<SubAgentOutput, BoxError> {
        let id = uuid::Uuid::new_v4().to_string();
        let name = "claude_code".to_owned();
        let push = |event: SubAgentEvent| {
            emit_subagent(SubAgentFrame {
                subagent_id: id.clone(),
                subagent_name: name.clone(),
                event,
            });
        };

        let mut cmd = Command::new(&self.config.claude_bin);
        cmd.args([
            "--print",
            "--output-format",
            "stream-json",
            "--verbose", // required by stream-json
            "--permission-mode",
            "bypassPermissions",
        ]);
        if let Some(m) = &self.config.model {
            cmd.args(["--model", m.as_str()]);
        }
        // Final positional: the task itself.
        cmd.arg(&input.task);
        // Run inside the sandbox root so `claude`'s default
        // workspace resolution lands in the right place.
        cmd.current_dir(&input.workspace_root)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true);

        push(SubAgentEvent::Started {
            task: input.task.clone(),
            model: self.config.model.clone(),
        });

        let mut child = cmd.spawn().map_err(|e| -> BoxError {
            format!("spawn `{}`: {e}", self.config.claude_bin).into()
        })?;

        let stdout = child
            .stdout
            .take()
            .ok_or_else(|| -> BoxError { "claude stdout missing".into() })?;
        let stderr = child
            .stderr
            .take()
            .ok_or_else(|| -> BoxError { "claude stderr missing".into() })?;

        // Drain stderr in the background so a chatty `claude` doesn't
        // deadlock against a full pipe. We surface its content only
        // when something goes wrong.
        let stderr_handle = tokio::spawn(async move {
            let mut reader = BufReader::new(stderr).lines();
            let mut buf = String::new();
            while let Ok(Some(line)) = reader.next_line().await {
                buf.push_str(&line);
                buf.push('\n');
                // Cap at 64 KiB to avoid unbounded memory if the CLI
                // logs verbose junk.
                if buf.len() > 64 * 1024 {
                    break;
                }
            }
            buf
        });

        let mut reader = BufReader::new(stdout).lines();
        let mut final_message = String::new();
        let mut error_message: Option<String> = None;
        // Track pending tool_use_id → name so we can pair tool_result
        // blocks back to the original tool name when emitting ToolEnd.
        let mut tool_names: HashMap<String, String> = HashMap::new();

        while let Some(line) = reader.next_line().await? {
            match parse_sdk_message(&line) {
                Some(SdkMessage::System { subtype, .. }) => {
                    // The init payload carries the model + tools list;
                    // not load-bearing for the UI, but a Status frame
                    // is the closest match.
                    if let Some(s) = subtype {
                        push(SubAgentEvent::Status {
                            message: format!("system:{s}"),
                        });
                    }
                }
                Some(SdkMessage::Assistant { message }) => {
                    for block in message.content {
                        match block {
                            ContentBlock::Text { text } => {
                                if !text.is_empty() {
                                    push(SubAgentEvent::Delta { text });
                                }
                            }
                            ContentBlock::ToolUse { id, name, input } => {
                                tool_names.insert(id, name.clone());
                                push(SubAgentEvent::ToolStart {
                                    name,
                                    arguments: input,
                                });
                            }
                            ContentBlock::Thinking { .. } | ContentBlock::Unknown => {
                                // Forward-compat: silently ignore.
                            }
                            ContentBlock::ToolResult { .. } => {
                                // Should only appear on user turns; ignore.
                            }
                        }
                    }
                }
                Some(SdkMessage::User { message }) => {
                    for block in message.content {
                        if let ContentBlock::ToolResult {
                            tool_use_id,
                            content,
                        } = block
                        {
                            let name = tool_names
                                .remove(&tool_use_id)
                                .unwrap_or_else(|| "unknown".into());
                            push(SubAgentEvent::ToolEnd {
                                name,
                                output: stringify_tool_result(content),
                            });
                        }
                    }
                }
                Some(SdkMessage::Result {
                    subtype,
                    result,
                    is_error,
                }) => {
                    if is_error || subtype.as_deref() != Some("success") {
                        error_message = Some(result.clone().unwrap_or_else(|| {
                            format!("claude returned error subtype={:?}", subtype)
                        }));
                    } else if let Some(r) = result {
                        final_message = r;
                    }
                }
                Some(SdkMessage::Unknown) | None => {
                    debug!(line = %line, "claude_code stream: skipping unparseable line");
                }
            }
        }

        let status = child.wait().await?;
        let stderr_text = stderr_handle.await.unwrap_or_default();

        if let Some(msg) = error_message {
            push(SubAgentEvent::Error {
                message: msg.clone(),
            });
            warn!(error = %msg, "claude_code error result");
            return Err(format!("subagent error: {msg}").into());
        }

        if !status.success() {
            let trimmed = stderr_text.trim();
            let msg = if trimmed.is_empty() {
                format!("`claude` exited {status}")
            } else {
                format!("`claude` exited {status}: {trimmed}")
            };
            push(SubAgentEvent::Error {
                message: msg.clone(),
            });
            return Err(msg.into());
        }

        push(SubAgentEvent::Done {
            final_message: final_message.clone(),
        });

        Ok(SubAgentOutput {
            message: final_message,
            artifacts: artifacts_for_run(),
        })
    }
}

/// Future enhancement: parse `git status` / `git diff` after the run
/// to populate `Artifact::FilesChanged`. v1.0 returns nothing; the
/// final-message text already names the changed files.
fn artifacts_for_run() -> Vec<Artifact> {
    Vec::new()
}

/// `claude --output-format stream-json` SDKMessage envelope.
/// Internal — the public-facing event surface is [`SubAgentEvent`].
#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum SdkMessage {
    System {
        #[serde(default)]
        subtype: Option<String>,
        #[serde(flatten)]
        _rest: serde_json::Map<String, Value>,
    },
    Assistant {
        message: AssistantInner,
    },
    User {
        message: UserInner,
    },
    Result {
        #[serde(default)]
        subtype: Option<String>,
        #[serde(default)]
        result: Option<String>,
        #[serde(default)]
        is_error: bool,
    },
    /// Forward-compat for new SDKMessage variants. Captured so we
    /// don't crash on an upstream addition; our match arm logs +
    /// drops them.
    #[serde(other)]
    Unknown,
}

#[derive(Debug, Deserialize)]
struct AssistantInner {
    #[serde(default)]
    content: Vec<ContentBlock>,
}

#[derive(Debug, Deserialize)]
struct UserInner {
    /// `content` is normally `Vec<ContentBlock>` for stream-json, but
    /// the CLI also emits plain-string content for some legacy paths.
    /// Default to empty if it's not an array.
    #[serde(default, deserialize_with = "deserialize_user_content")]
    content: Vec<ContentBlock>,
}

fn deserialize_user_content<'de, D>(d: D) -> Result<Vec<ContentBlock>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    use serde::de::Error;
    let v = Value::deserialize(d)?;
    match v {
        Value::Array(_) => serde_json::from_value(v).map_err(D::Error::custom),
        _ => Ok(Vec::new()),
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
enum ContentBlock {
    Text {
        #[serde(default)]
        text: String,
    },
    ToolUse {
        id: String,
        name: String,
        #[serde(default)]
        input: Value,
    },
    ToolResult {
        tool_use_id: String,
        /// Can be a string OR an array of `{type:"text",text}` blocks.
        /// We keep it as a `Value` and flatten via
        /// [`stringify_tool_result`] when emitting to the UI.
        #[serde(default)]
        content: Value,
    },
    /// Claude's extended-thinking surface. We recognise the
    /// variant so it doesn't fall into `Unknown` (and pollute the
    /// debug log), but the body is intentionally ignored — the
    /// inline card already shows assistant text + tool calls and
    /// dumping reasoning verbatim adds noise without helping the
    /// reader follow the loop.
    #[allow(dead_code)]
    Thinking {
        #[serde(default)]
        thinking: String,
    },
    #[serde(other)]
    Unknown,
}

fn parse_sdk_message(line: &str) -> Option<SdkMessage> {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return None;
    }
    serde_json::from_str(trimmed).ok()
}

/// Tool results land as either a plain string or an array of
/// `{type:"text",text}` blocks. Flatten to a single string so the
/// `SubAgentEvent::ToolEnd.content` field stays simple.
fn stringify_tool_result(v: Value) -> String {
    match v {
        Value::String(s) => s,
        Value::Array(items) => {
            let mut out = String::new();
            for item in items {
                if let Some(text) = item.get("text").and_then(|t| t.as_str()) {
                    if !out.is_empty() {
                        out.push('\n');
                    }
                    out.push_str(text);
                }
            }
            out
        }
        Value::Null => String::new(),
        other => other.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_assistant_text_block() {
        let line = r#"{"type":"assistant","message":{"content":[{"type":"text","text":"hello"}]}}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::Assistant { message } => match &message.content[0] {
                ContentBlock::Text { text } => assert_eq!(text, "hello"),
                other => panic!("unexpected block: {other:?}"),
            },
            other => panic!("unexpected msg: {other:?}"),
        }
    }

    #[test]
    fn parses_assistant_tool_use() {
        let line = r#"{"type":"assistant","message":{"content":[{"type":"tool_use","id":"u1","name":"Read","input":{"path":"a.rs"}}]}}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::Assistant { message } => match &message.content[0] {
                ContentBlock::ToolUse { id, name, input } => {
                    assert_eq!(id, "u1");
                    assert_eq!(name, "Read");
                    assert_eq!(input["path"], "a.rs");
                }
                other => panic!("unexpected block: {other:?}"),
            },
            other => panic!("unexpected msg: {other:?}"),
        }
    }

    #[test]
    fn parses_user_tool_result_string_content() {
        let line = r#"{"type":"user","message":{"content":[{"type":"tool_result","tool_use_id":"u1","content":"file body"}]}}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::User { message } => match &message.content[0] {
                ContentBlock::ToolResult {
                    tool_use_id,
                    content,
                } => {
                    assert_eq!(tool_use_id, "u1");
                    assert_eq!(stringify_tool_result(content.clone()), "file body");
                }
                other => panic!("unexpected block: {other:?}"),
            },
            other => panic!("unexpected msg: {other:?}"),
        }
    }

    #[test]
    fn parses_user_tool_result_array_content() {
        let line = r#"{"type":"user","message":{"content":[{"type":"tool_result","tool_use_id":"u1","content":[{"type":"text","text":"line a"},{"type":"text","text":"line b"}]}]}}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::User { message } => match &message.content[0] {
                ContentBlock::ToolResult { content, .. } => {
                    assert_eq!(stringify_tool_result(content.clone()), "line a\nline b");
                }
                other => panic!("unexpected block: {other:?}"),
            },
            other => panic!("unexpected msg: {other:?}"),
        }
    }

    #[test]
    fn parses_result_success() {
        let line = r#"{"type":"result","subtype":"success","result":"all done","is_error":false}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::Result {
                subtype,
                result,
                is_error,
            } => {
                assert_eq!(subtype.as_deref(), Some("success"));
                assert_eq!(result.as_deref(), Some("all done"));
                assert!(!is_error);
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn parses_result_error() {
        let line = r#"{"type":"result","subtype":"error_max_turns","is_error":true}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::Result {
                subtype, is_error, ..
            } => {
                assert_eq!(subtype.as_deref(), Some("error_max_turns"));
                assert!(is_error);
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn unknown_message_kind_does_not_panic() {
        let line = r#"{"type":"future_kind_we_havent_seen","stuff":1}"#;
        match parse_sdk_message(line).unwrap() {
            SdkMessage::Unknown => {}
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn skips_blank_and_garbage_lines() {
        assert!(parse_sdk_message("").is_none());
        assert!(parse_sdk_message("   ").is_none());
        assert!(parse_sdk_message("not json").is_none());
    }

    #[test]
    fn stringify_handles_null_and_unknown() {
        assert_eq!(stringify_tool_result(Value::Null), "");
        assert_eq!(stringify_tool_result(Value::Bool(true)), "true");
    }

    #[tokio::test]
    async fn probe_returns_err_when_binary_missing() {
        let err = probe("/no/such/binary-jarvis-test").await.unwrap_err();
        assert!(err.contains("spawn"), "unexpected: {err}");
    }
}

File: packages/subagents/src/claude-code.ts
// `subagent.claude_code` — delegate a coding task to the local `claude` CLI
// (Anthropic's Claude Code distribution). Spawns
// `claude --print --output-format stream-json --verbose --permission-mode
// bypassPermissions [--model M] [extra...] <task>`, parses each SDKMessage
// line on stdout, and emits the equivalent SubAgentEvent frames. Ported from
// harness-subagents/src/claude_code.rs.
//
// Discovery: the composition root calls `probe` once at startup; if `claude`
// isn't on PATH (or --version fails / is unrecognised) it rejects and the
// subagent is NOT registered (no panic).
//
// Authentication: the spawned `claude` inherits the parent process env, so an
// interactively-logged-in user gets seamless auth. Extra CLI flags (e.g.
// `--bare` in keychain-less environments) are forwarded via
// `ClaudeCodeConfig.extra_args`.
import { spawn } from "node:child_process";
import { execFile } from "node:child_process";
import { createInterface } from "node:readline";
import type { JsonValue } from "@jarvis/core";
import {
  emitSubagent,
  type Artifact,
  type SubAgent,
  type SubAgentEvent,
  type SubAgentInput,
  type SubAgentOutput,
} from "./subagent.ts";

export const DESCRIPTION =
  "Delegate a coding task to Claude Code (Anthropic's `claude` CLI). It can read, edit, and run shell commands inside the workspace; treat it like a peer coder you hand a focused task to. Will modify files; the call is approval-gated.";

export interface ClaudeCodeConfig {
  /** Path or name of the `claude` binary. Default `claude` (PATH lookup). */
  claude_bin: string;
  /** Optional model override forwarded as `--model`. `undefined` = CLI default. */
  model?: string;
  /**
   * Extra args forwarded verbatim, inserted after the standard flags + any
   * `--model`, before the final task positional. Empty by default. Absent on
   * the wire when empty (serde `default`).
   */
  extra_args?: string[];
  /**
   * Wall-clock budget for a single run, in milliseconds. When the spawned
   * `claude` outruns it the child is SIGKILL'd and `invoke` rejects — a hung
   * CLI must never block the caller forever. Defaults to
   * {@link CLAUDE_CODE_DEFAULT_TIMEOUT_MS}.
   */
  timeout_ms?: number;
}

/**
 * Default run timeout: 10 minutes, matching `JARVIS_WORK_RUN_TIMEOUT_MS`
 * (the auto-loop's per-run budget) so a delegated `claude` run can't outlast
 * the work unit that dispatched it.
 */
export const CLAUDE_CODE_DEFAULT_TIMEOUT_MS = 600_000;

export function defaultClaudeCodeConfig(): ClaudeCodeConfig {
  return { claude_bin: "claude" };
}

/**
 * Probe the host: is `claude` on PATH and does `--version` recognise it as
 * Claude Code? Run once at startup; on failure callers skip registration.
 * Rejects with a reason string (mirrors Rust's `Result<(), String>`).
 */
export function probe(claudeBin: string): Promise<void> {
  return new Promise((resolve, reject) => {
    execFile(claudeBin, ["--version"], { encoding: "utf8" }, (err, stdout) => {
      if (err) {
        // ENOENT (binary missing) or non-zero exit both land here.
        if ((err as { code?: string }).code === "ENOENT") {
          reject(new Error(`spawn \`${claudeBin}\` failed: ${err.message}`));
        } else {
          reject(new Error(`\`${claudeBin} --version\` exited with error: ${err.message}`));
        }
        return;
      }
      // The CLI prints e.g. `2.1.119 (Claude Code)` — the parenthesised tag
      // disambiguates the official binary from someone else's `claude`.
      if (!stdout.includes("Claude Code")) {
        reject(new Error(`\`${claudeBin} --version\` output unrecognised (got: ${stdout.trim()})`));
        return;
      }
      resolve();
    });
  });
}

const STDERR_CAP = 64 * 1024;

export class ClaudeCodeSubAgent implements SubAgent {
  readonly name = "claude_code";
  readonly description = DESCRIPTION;
  readonly #config: ClaudeCodeConfig;

  constructor(config: ClaudeCodeConfig) {
    this.#config = config;
  }

  inputSchema(): JsonValue {
    return {
      type: "object",
      properties: {
        task: {
          type: "string",
          description: "Natural-language coding task. Claude Code will plan + execute autonomously.",
        },
      },
      required: ["task"],
      additionalProperties: false,
    };
  }

  requiresApproval(): boolean {
    // The CLI will mutate the workspace. The outer approver MUST gate this;
    // per-tool approvals inside the CLI run under bypassPermissions (the user
    // already approved the outer call, and --print mode can't surface prompts).
    return true;
  }

  async invoke(input: SubAgentInput): Promise<SubAgentOutput> {
    const id = crypto.randomUUID();
    const push = (event: SubAgentEvent): void => {
      emitSubagent({ subagent_id: id, subagent_name: this.name, event });
    };

    const args = ["--print", "--output-format", "stream-json", "--verbose", "--permission-mode", "bypassPermissions"];
    if (this.#config.model !== undefined) args.push("--model", this.#config.model);
    if (this.#config.extra_args && this.#config.extra_args.length > 0) args.push(...this.#config.extra_args);
    args.push(input.task); // final positional

    const startEvent: SubAgentEvent = { kind: "started", task: input.task };
    if (this.#config.model !== undefined) startEvent.model = this.#config.model;
    push(startEvent);

    let child;
    try {
      child = spawn(this.#config.claude_bin, args, {
        cwd: input.workspace_root,
        stdio: ["ignore", "pipe", "pipe"],
      });
    } catch (e) {
      throw new Error(`spawn \`${this.#config.claude_bin}\`: ${stringifyError(e)}`);
    }

    // A spawn failure (ENOENT) surfaces asynchronously via the 'error' event,
    // not as a throw — capture it so `invoke` rejects rather than hanging.
    let spawnError: Error | undefined;
    child.on("error", (e) => {
      spawnError = e;
    });

    // Drain stderr in the background, capped at 64 KiB. Swallow stream errors
    // on the diagnostic channel so an stderr EPIPE can't escape as an
    // unhandled rejection (the run's success is judged by the result message
    // and exit code, not by stderr).
    let stderrText = "";
    if (child.stderr) {
      child.stderr.setEncoding("utf8");
      child.stderr.on("data", (chunk: string) => {
        if (stderrText.length < STDERR_CAP) stderrText += chunk;
      });
      child.stderr.on("error", () => {});
    }

    let finalMessage = "";
    let errorMessage: string | undefined;
    // tool_use_id -> tool name, so tool_result blocks pair back to ToolEnd.
    const toolNames = new Map<string, string>();

    // Enforce a wall-clock budget: a hung/stalled CLI is SIGKILL'd so `invoke`
    // can never await forever. The kill closes stdout, ending the loop below,
    // and the flag is surfaced as a timeout error after the child is reaped.
    const timeoutMs = this.#config.timeout_ms ?? CLAUDE_CODE_DEFAULT_TIMEOUT_MS;
    let timedOut = false;
    let streamError: Error | undefined;
    const timer = setTimeout(() => {
      timedOut = true;
      child.kill("SIGKILL");
    }, timeoutMs);

    try {
      if (child.stdout) {
        child.stdout.setEncoding("utf8");
        // Capture an stdout stream error (e.g. EPIPE) rather than letting it
        // reject the `for await` and unwind out of `invoke` with the child
        // still alive — the `finally` below reaps it regardless.
        child.stdout.on("error", (e: Error) => {
          if (streamError === undefined) streamError = e;
        });
        const lines = createInterface({ input: child.stdout, crlfDelay: Infinity });
        for await (const line of lines) {
          const msg = parseSdkMessage(line);
          if (msg === undefined) continue; // blank / garbage
          if (msg.type === "system") {
            if (msg.subtype !== undefined) push({ kind: "status", message: `system:${msg.subtype}` });
          } else if (msg.type === "assistant") {
            for (const block of msg.message.content) {
              if (block.type === "text") {
                if (block.text.length > 0) push({ kind: "delta", text: block.text });
              } else if (block.type === "tool_use") {
                toolNames.set(block.id, block.name);
                push({ kind: "tool_start", name: block.name, arguments: block.input });
              }
              // thinking / tool_result on assistant turns: forward-compat ignore.
            }
          } else if (msg.type === "user") {
            for (const block of msg.message.content) {
              if (block.type === "tool_result") {
                const name = toolNames.get(block.tool_use_id) ?? "unknown";
                toolNames.delete(block.tool_use_id);
                push({ kind: "tool_end", name, output: stringifyToolResult(block.content) });
              }
            }
          } else if (msg.type === "result") {
            if (msg.is_error || msg.subtype !== "success") {
              errorMessage = msg.result ?? `claude returned error subtype=${JSON.stringify(msg.subtype)}`;
            } else if (msg.result !== undefined) {
              finalMessage = msg.result;
            }
          }
          // unknown message kind: skip silently.
        }
      }
    } catch (e) {
      // A throw from the stream (or the loop body) must not orphan the child.
      if (streamError === undefined) streamError = e instanceof Error ? e : new Error(String(e));
    } finally {
      // Reap the child on any abnormal exit path — Node does not kill children
      // on GC, so an un-reaped `claude` (running under bypassPermissions) would
      // keep mutating the workspace and burning quota. On a clean completion we
      // leave it to exit on its own so the real exit code is observed below.
      if ((timedOut || streamError !== undefined) && child.exitCode === null && child.signalCode === null) {
        child.kill("SIGKILL");
      }
    }

    const code = await waitForExit(child);
    clearTimeout(timer);

    if (timedOut) {
      const msg = `\`claude\` timed out after ${timeoutMs} ms`;
      push({ kind: "error", message: msg });
      throw new Error(msg);
    }
    if (spawnError !== undefined) {
      throw new Error(`spawn \`${this.#config.claude_bin}\`: ${spawnError.message}`);
    }
    if (streamError !== undefined) {
      const msg = `\`claude\` stdout stream error: ${streamError.message}`;
      push({ kind: "error", message: msg });
      throw new Error(msg);
    }

    if (errorMessage !== undefined) {
      push({ kind: "error", message: errorMessage });
      throw new Error(`subagent error: ${errorMessage}`);
    }

    if (code !== 0) {
      const trimmed = stderrText.trim();
      const msg = trimmed.length === 0 ? `\`claude\` exited ${code}` : `\`claude\` exited ${code}: ${trimmed}`;
      push({ kind: "error", message: msg });
      throw new Error(msg);
    }

    push({ kind: "done", final_message: finalMessage });

    const artifacts = artifactsForRun();
    const out: SubAgentOutput = { message: finalMessage };
    if (artifacts.length > 0) out.artifacts = artifacts;
    return out;
  }
}

/** Future enhancement: parse git status/diff to populate FilesChanged. */
function artifactsForRun(): Artifact[] {
  return [];
}

// ---------------------------------------------------------------------------
// SDKMessage envelope parsing (claude --output-format stream-json)
// ---------------------------------------------------------------------------

type ContentBlock =
  | { type: "text"; text: string }
  | { type: "tool_use"; id: string; name: string; input: JsonValue }
  | { type: "tool_result"; tool_use_id: string; content: JsonValue }
  | { type: "thinking"; thinking: string }
  | { type: "__unknown__" };

type SdkMessage =
  | { type: "system"; subtype?: string }
  | { type: "assistant"; message: { content: ContentBlock[] } }
  | { type: "user"; message: { content: ContentBlock[] } }
  | { type: "result"; subtype?: string; result?: string; is_error: boolean }
  | { type: "__unknown__" };

/** Parse one stdout line into an SDKMessage; undefined for blank/garbage. */
export function parseSdkMessage(line: string): SdkMessage | undefined {
  const trimmed = line.trim();
  if (trimmed.length === 0) return undefined;
  let raw: unknown;
  try {
    raw = JSON.parse(trimmed);
  } catch {
    return undefined;
  }
  if (typeof raw !== "object" || raw === null) return undefined;
  const obj = raw as Record<string, unknown>;
  switch (obj["type"]) {
    case "system":
      return { type: "system", subtype: typeof obj["subtype"] === "string" ? obj["subtype"] : undefined };
    case "assistant":
      return { type: "assistant", message: { content: parseContent(messageContent(obj)) } };
    case "user":
      // `content` is normally an array for stream-json; non-arrays (legacy
      // plain-string content) default to empty.
      return { type: "user", message: { content: parseContent(messageContent(obj)) } };
    case "result":
      return {
        type: "result",
        subtype: typeof obj["subtype"] === "string" ? obj["subtype"] : undefined,
        result: typeof obj["result"] === "string" ? obj["result"] : undefined,
        is_error: obj["is_error"] === true,
      };
    default:
      return { type: "__unknown__" };
  }
}

function messageContent(obj: Record<string, unknown>): unknown {
  const message = obj["message"];
  if (typeof message !== "object" || message === null) return undefined;
  return (message as Record<string, unknown>)["content"];
}

function parseContent(content: unknown): ContentBlock[] {
  if (!Array.isArray(content)) return [];
  const out: ContentBlock[] = [];
  for (const item of content) {
    if (typeof item !== "object" || item === null) continue;
    const block = item as Record<string, unknown>;
    switch (block["type"]) {
      case "text":
        out.push({ type: "text", text: typeof block["text"] === "string" ? block["text"] : "" });
        break;
      case "tool_use":
        out.push({
          type: "tool_use",
          id: typeof block["id"] === "string" ? block["id"] : "",
          name: typeof block["name"] === "string" ? block["name"] : "",
          input: (block["input"] ?? {}) as JsonValue,
        });
        break;
      case "tool_result":
        out.push({
          type: "tool_result",
          tool_use_id: typeof block["tool_use_id"] === "string" ? block["tool_use_id"] : "",
          content: (block["content"] ?? null) as JsonValue,
        });
        break;
      case "thinking":
        out.push({ type: "thinking", thinking: typeof block["thinking"] === "string" ? block["thinking"] : "" });
        break;
      default:
        out.push({ type: "__unknown__" });
    }
  }
  return out;
}

/**
 * Tool results land as either a plain string or an array of `{type,text}`
 * blocks. Flatten to a single string. Null -> "". Other -> JSON.
 */
export function stringifyToolResult(v: JsonValue): string {
  if (typeof v === "string") return v;
  if (v === null) return "";
  if (Array.isArray(v)) {
    let out = "";
    for (const item of v) {
      if (typeof item === "object" && item !== null && !Array.isArray(item)) {
        const text = (item as Record<string, JsonValue>)["text"];
        if (typeof text === "string") {
          if (out.length > 0) out += "\n";
          out += text;
        }
      }
    }
    return out;
  }
  return JSON.stringify(v);
}

function waitForExit(child: ReturnType<typeof spawn>): Promise<number> {
  return new Promise((resolve) => {
    child.on("close", (code) => resolve(code ?? 0));
  });
}

function stringifyError(e: unknown): string {
  if (e instanceof Error) return e.message;
  return String(e);
}
